docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+```csharp
+var transform = context.Sources.AdditionalTexts
+                               .Transform(t => t.Path)
+                               .Transform(p => "prefix_" + p);
+```
+
+During the first execution of the pipeline each of the two lambdas will be
+executed for each additional file:
+
+AdditionalText          | Transform1 | Transform2
+------------------------|------------|-----------------
+Text{ Path: "abc.txt" } | "abc.txt"  | "prefix_abc.txt"
+Text{ Path: "def.txt" } | "def.txt"  | "prefix_def.txt"
+Text{ Path: "ghi.txt" } | "ghi.txt"  | "prefix_ghi.txt"
+
+Now consider the case where in some future iteration, the first additional file
+has changed and has a different path, and the second file has changed, but kept
+its path the same.
+
+AdditionalText               | Transform1 | Transform2
+-----------------------------|------------|-----------
+**Text{ Path: "diff.txt" }** |            |
+**Text{ Path: "def.txt" }**  |            |
+Text{ Path: "ghi.txt" }      |            |
+
+The generator would run transform1 on the first and second files, producing
+"diff.txt" and "def.txt" respectively. However, it would not need to re-run the
+transform for the third file, as the input has not changed. It can just use the
+previously cached value.
+
+AdditionalText               | Transform1     | Transform2
+-----------------------------|----------------|-----------
+**Text{ Path: "diff.txt" }** | **"diff.txt"** |
+**Text{ Path: "def.txt" }**  | **"def.txt"**  |
+Text{ Path: "ghi.txt" }      | "ghi.txt"      |
+
+Next the driver would look to run Transform2. It would operate on `"diff.txt"`
+producing `"prefix_diff.txt"`, but when it comes to `"def.txt"` it can observe
+that the item produced was the same as the last iteration. Even though the
+original input (`Text{ Path: "def.txt" }`) was changed, the result of Transform1
+on it was the same. Thus there is no need to re-run Transform2 on `"def.txt"` as
+it can just use the cached value from before. Similarly the cached state of
+"ghi.txt" can be used.
+
+AdditionalText               | Transform1     | Transform2
+-----------------------------|----------------|----------------------
+**Text{ Path: "diff.txt" }** | **"diff.txt"** | **"prefix_diff.txt"**
+**Text{ Path: "def.txt" }**  | **"def.txt"**  | "prefix_diff.txt"
+Text{ Path: "ghi.txt" }      | "ghi.txt"      | "prefix_ghi.txt"
+
+In this way, only changes that are consequential flow through the pipeline, and
+duplicate work is avoided. If a generator only relies on `AdditionalTexts` then
+the driver knows there can be no work to be done when a `SyntaxTree` changes.
+
+### WithComparer
+
+For a user provided result to be comparable across iterations, there needs to be some concept of equivalence. Rather than requiring types returned from transformations implement `IEquatable<T>`, there exists an extension method that allows the author to supply a comparer that should be used when comparing values for the given transformation:
+
```

Comment:
do we specify what semantics you get without this?  is it EqualityComparer<T>.Default?